specification/provider/providers.md

```diff
@@ -1,5 +1,159 @@
+>
+> ##### Conditional Requirement 2.3.1
+>
+> > The `feature provider` interface **MUST** define methods for typed flag resolution, including boolean, numeric, string, and structure.
+
+```
+// example boolean flag value resolution
+FlagResolution resolveBooleanValue(string flagKey, boolean defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example string flag value resolution
+FlagResolution resolveStringValue(string flagKey, string defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example number flag value resolution
+FlagResolution resolveNumberValue(string flagKey, number defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example structure flag value resolution
+FlagResolution resolveStructureValue(string flagKey, JsonObject defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+```
+
+##### Requirement 2.4
+
+> In cases of normal execution, the `provider` **MUST** populate the `flag resolution` structure's `value` field with the resolved flag value.
+
+##### Requirement 2.4
+
+> In cases of abnormal execution, the `provider` **MUST** populate the `flag resolution` structure's `value` field with the default flag value.
+
+##### Requirement 2.5
+
+> In cases of normal execution, the `provider` **SHOULD** populate the `flag resolution` structure's `variant` field with a string identifier corresponding to the returned flag value.
+
+For example, the flag value might be `3.14159265359`, and the variant field's value might be `"pi"`.
+
+The value of the variant field might only be meaningful in the context of the flag management system associated with the provider. For example, the variant may be a UUID corresponding to the variant in the flag management system, or an index corresponding to the variant in the flag management system.
+
+##### Requirement 2.6
+
+> The `provider` **SHOULD** populate the `flag resolution` structure's `reason` field with a string indicating the semantic reason for the returned flag value.
+
+Possible values vary by provider, but might include such values as `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` or `"ERROR"`.
+
+##### Requirement 2.7
+
+> In cases of normal execution, the `provider` **MUST NOT** populate the `flag resolution` structure's `error code` field, or otherwise must populate it with a null or falsy value.
+
+##### Requirement 2.8
+
+> In cases of abnormal execution, the `provider` **MUST** indicate an error using the idioms of the implementation language, with an associated error code having possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.
+
+The provider might throw an exception, return an error, or populate the `error code` object on the returned `flag resolution` structure to indicate a problem during flag value resolution.
```

Comment:
I thought that we didn't want the provider throwing an error?

Reply:
Most of the sample providers throw errors, consistent with the "throw early, catch late" mantra. They are all caught in the client, which calls the providers (providers are not directly accessed by application authors), and must never throw.

Reply:
I think it might be worth mentioning that here - "a provider might throw an exception (which will always be caught within the client and not surfaced to application authors)" or something.